sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -0,0 +1,253 @@
+        List<OpenConnectionTask> openConnectionTasks = this.endpointsUnderMonitorMap.remove(addressUriString);
+
+        if (openConnectionTasks != null && !openConnectionTasks.isEmpty()) {
+            for (OpenConnectionTask connectionTask : openConnectionTasks) {
+                connectionTask.complete(openConnectionResponse);
+            }
+        }
+    }
+
+    private void reInstantiateOpenConnectionsPublisherAndSubscribe() {
+        logger.debug("In open connections task sink and concurrency reduction flow");
+        this.forceDefensiveOpenConnections();
+        this.instantiateOpenConnectionsPublisher();
+        this.openConnectionBackgroundTask.dispose();
+        this.openConnectionBackgroundTask = this.getOpenConnectionsPublisher();
+    }
+
+    private Mono<OpenConnectionResponse> enqueueOpenConnectionOpsForRetry(
+            OpenConnectionTask op,
+            ShouldRetryResult retryResult) {
+        if (retryResult.backOffTime == Duration.ZERO || retryResult.backOffTime == null) {
+            this.submitOpenConnectionWithinLoopInternal(op);
+            return Mono.empty();
+        } else {
+            return Mono
+                    .delay(retryResult.backOffTime)
+                    .flatMap(ignore -> {
+                        this.submitOpenConnectionWithinLoopInternal(op);
+                        return Mono.empty();
+                    });
+        }
+    }
+
+    // when the flux associated with openConnectionsTaskSink is cancelled
+    // this method provides for a way to resume emitting pushed tasks to
+    // the sink
+    private synchronized void instantiateOpenConnectionsPublisher() {
+        logger.debug("Re-instantiate open connections task sink");
+        openConnectionsTaskSink = openConnectionsTaskSinkBackUp;
+        openConnectionsTaskSinkBackUp = Sinks.many().multicast().onBackpressureBuffer();
+    }
+
+    private void forceDefensiveOpenConnections() {
+        if (aggressivenessHint.get() == AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE) {
+            aggressivenessHint.set(AsyncDocumentClient.OpenConnectionAggressivenessHint.DEFENSIVE);
+        }
+    }
+
+    private void completeSink(Sinks.Many<OpenConnectionTask> sink) {
+        Sinks.EmitResult completeEmitResult = sink.tryEmitComplete();
+
+        if (completeEmitResult == Sinks.EmitResult.OK) {
+            logger.debug("Sink completed.");
+        } else if (completeEmitResult == Sinks.EmitResult.FAIL_CANCELLED ||
+                completeEmitResult == Sinks.EmitResult.FAIL_TERMINATED) {
+            logger.debug("Sink already completed, EmitResult: {}", completeEmitResult);
+        } else {
+            logger.warn("Sink completion failed, EmitResult: {}", completeEmitResult);
+        }
+    }
+
+    private static class ConcurrencyConfiguration {
+        int openConnectionOperationEmissionConcurrency;
+        int openConnectionExecutionConcurrency;
```

Comment:
Both of these should be final. Always a good practice to have immutable objects, so you don't have to worry about state and your code is stateless.